olp-cpp-sdk-core/src/network2/winhttp/NetworkProtocolWinHttp.cpp

```diff
@@ -0,0 +1,1063 @@
+            if (handle->payload->tellp() !=
+                std::streampos(handle->result_data->count)) {
+              handle->payload->seekp(handle->result_data->count);
+              if (handle->payload->fail()) {
+                EDGE_SDK_LOG_WARNING(
+                    LOGTAG,
+                    "Reception stream doesn't support setting write point");
+                handle->payload->clear();
+              }
+            }
+            //}
+
+            handle->payload->write(dataBuffer, dataLen);
+          }
+          handle->result_data->count += dataLen;
+        }
+      }
+      LocalFree((HLOCAL)dataBuffer);
+    }
+
+    WinHttpQueryDataAvailable(handle->http_request, NULL);
+  } else if (status == WINHTTP_CALLBACK_STATUS_HANDLE_CLOSING) {
+    // Only now is it safe to free the handle
+    // See
+    // https://docs.microsoft.com/en-us/windows/desktop/api/winhttp/nf-winhttp-winhttpclosehandle
+    handle->freeHandle();
+    return;
+  } else {
+    EDGE_SDK_LOG_ERROR(LOGTAG,
+                       "Unknown callback " << std::hex << status << std::dec);
+  }
+}
+
+DWORD
+NetworkProtocolWinHttp::Run(LPVOID arg) {
+  reinterpret_cast<NetworkProtocolWinHttp*>(arg)->CompletionThread();
+  return 0;
+}
+
+void NetworkProtocolWinHttp::CompletionThread() {
+  std::shared_ptr<NetworkProtocolWinHttp> that = shared_from_this();
+
+  while (that->http_session_) {
+    std::shared_ptr<ResultData> result;
+    {
+      if (http_session_ && results_.empty()) {
+        WaitForSingleObject(event_, 30000);  // Wait max 30 seconds
+        ResetEvent(event_);
+      }
+      if (!http_session_) continue;
+
+      std::unique_lock<std::recursive_mutex> lock(mutex_);
+      if (!results_.empty()) {
+        result = results_.front();
+        results_.pop();
+      }
+    }
+
+    if (http_session_ && result) {
+      std::string str;
+      int status;
+      if ((result->offset == 0) && (result->status == 206))
+        result->status = 200;
+
+      if (result->completed)
+        str =
+            "todo: convert http code to string";  // NetworkProtocol::HttpErrorToString(result->status);
+      else
+        str = errorToString(result->status);
+
+      if (result->completed)
+        status = result->status;
+      else
+        status = static_cast<int>(errorToCode(result->status));
+      // NetworkResponse response(
+      //    result->id, result->cancelled, status, str, result->maxAge,
+      //    result->expires, result->etag, result->contentType,
+      //    result->count, result->offset, result->payload,
+      //    std::move(result->statistics));
+      NetworkResponse response = NetworkResponse()
+                                     .WithCancelled(result->cancelled)
+                                     .WithError(str)
+                                     .WithRequestId(result->request_id)
+                                     .WithStatus(status);
+      if (result->user_callback) {
```

Comment:
maybe optimize and not even create NetworkResponse if nobody is waiting for an answer